### ISSUE TITLE
IPFS support

### ISSUE THREAD

**Issue by JayBrown** [lhc70000/iina#806]
* feature request
* any system
* any version

You can open video and audio files hosted on the IPFS with IINA, e.g. this one:

https://ipfs.io/ipfs/QmX8LDhDSYdX3xG6cHFUybXLDSuvo9Lz6wF5NU3UVmJRnB

However, this still relies on a working http gateway like ipfs.io. So it would be awesome to have IINA include an in-app ipfs node, running its own daemon, with its own ipfs data directory, and the user would then just haveh to input the IPFS hash, in this case `QmX8LDhDSYdX3xG6cHFUybXLDSuvo9Lz6wF5NU3UVmJRnB`.

More info on IPFS:
go-ipfs: https://github.com/ipfs/go-ipfs
js-ipfs: https://github.com/ipfs/js-ipfs

**Comment by skyline75489** [comment]
IPFS seems a really cool idea. But I don't think users would like to join the network unknowingly. Adding an option would be more acceptable for me.

**Comment by JayBrown** [comment]
I strongly beg to differ: supporting IPFS means joining the network. The daemon needs to be running anyway, so the question is just to what extent an IINA user will join the network.

(1) The standard way would be the IINA equivalent of `ipfs cat <hash> | mpv -`. In this case the media file is cached by the IINA node temporarily, seeding it for other users as long as it's cached. That's what the IPFS is all about, making content available across the network from many nodes. That must be the **default setting**, if there is to be "IPFS support". When IINA is quit, then you clear the cache, and the file won't be seeded anymore at next launch. (Or you mark the object as "to be deleted", and the cache is cleared at next launch, so you don't need a background helper process.)

(2) However, the user should have the ability to **opt-out of the default setting**, e.g. if he wants to consume pirated content and is paranoid about PopcornTime-like repercussions, and then IINA would just do the equivalent to `ipfs get <hash> -o /TempFolder/MediaFile`, playing back the local file from /TempFolder, and deleting it at application quit; no seeding occurs in this setup. (That's basically like entering a full gateway URL as above, the one with "ipfs.io".)

(3) As a second option the user could **opt-in and enable media pinning**. This would be the IINA equivalent of `ipfs pin <hash>`; then the media object would not be cached temporarily, but pinned & seeded indefinitely, until the user removes it again. From a coding perspective this would be the most complicated to add, because the user would need a window, where he can see his IPFS media pool, i.e. how many media files he has pinned, and from which he can remove the ones he doesn't want to seed anymore.

So that's what IPFS support would mean. Basic caching as default, opt-out to playback-only, opt-in to pin & seed.

**Comment by skyline75489** [comment]
@JayBrown Agreed. All I care about is basically the ability to opt-out. However, IPFS itself seems to be in heavy development in their own words. And the only complete implementation is in Golang. Do you have any successful example that integrate with IPFS on macOS ?

**Comment by JayBrown** [comment]
People usually only talk about go & javascript, when they talk IPFS. But there's also a **[Swift library](https://github.com/ipfs/swift-ipfs-api)**, which is what IINA would need, right? You could, of course, also add the go or js version of ipfs into ./Contents/MacOS, as there's also `youtube-dl` in there already, and somehow work with those executables, but that's probably not a very elegant solution. ;)
